llmfoundry/models/hf/hf_causal_lm.py

```diff
@@ -76,197 +114,181 @@ def __init__(self, om_model_config: Union[DictConfig,
-
-            # If the HuggingFace model is coming from a local folder, Hugging Face copies the modules into the
-            # transformers modules cache. On particular systems, this operation seems to cause contention between
-            # the different processes. To avoid this contention, we first create the model (on meta device) on local rank
-            # zero. This will set up the transformers model cache and avoid the future contention.
-            if dist.get_local_rank() == 0 and os.path.isdir(
-                    om_model_config.pretrained_model_name_or_path):
-                with init_empty_weights(include_buffers=False):
-                    with warnings.catch_warnings():
-                        warnings.simplefilter('ignore', UserWarning)
-                        AutoModelForCausalLM.from_pretrained(
-                            om_model_config.pretrained_model_name_or_path,
-                            trust_remote_code=trust_remote_code,
-                            use_auth_token=use_auth_token,
-                            config=config,
-                        )
-
-            dist.barrier()
-
-            # initialize the model on the correct device
-            if resolved_init_device == 'cpu':
-                if om_model_config.pretrained:
-                    model = AutoModelForCausalLM.from_pretrained(
+                        f'Config dict override got unknown keys. ' +
+                        f'Extra keys: {extra_keys}. ' +
+                        f'Expected (a subset of) keys: {list(attr.keys())}.')
+                getattr(config, k).update(v)
+            # necessary case to allow for rope_scaling to be overriden in llama config
+            elif attr is None and isinstance(v, Mapping):
+                setattr(config, k, {})
+                getattr(config, k).update(v)
+            else:
+                setattr(config, k, v)
+
+        # We need to have all non-zero local ranks be not-pretrained
+        # Rank 0 will still be pretrained, and distribute the weights appropriately
+        if dist.get_local_rank() != 0 and init_device == 'mixed':
+            om_model_config.pretrained = False
+
+        # If the HuggingFace model is coming from a local folder, Hugging Face copies the modules into the
+        # transformers modules cache. On particular systems, this operation seems to cause contention between
+        # the different processes. To avoid this contention, we first create the model (on meta device) on local rank
+        # zero. This will set up the transformers model cache and avoid the future contention.
+        if dist.get_local_rank() == 0 and os.path.isdir(
+                om_model_config.pretrained_model_name_or_path):
+            with init_empty_weights(include_buffers=False):
+                with warnings.catch_warnings():
+                    warnings.simplefilter('ignore', UserWarning)
+                    AutoModelForCausalLM.from_pretrained(
```

Comment:
should this warning filter be added for other `AutoModelForCausalLM.from_pretrained` callsites? if so, a helper to do this would be nice.

Reply:
No, this is just for this "pre" creation on meta device, cause it will raise a bunch of (generally correct) warnings that you're trying to load a checkpoint into a model with meta tensors which is probably not what you intended.